src/main/java/com/gtnewhorizons/modularui/api/math/MathExpression.java

```diff
@@ -1,315 +1,593 @@
+        return stack.get(0).value;
+    }
 
-        Object prevToken = null;
-        Object thisToken = null;
-        for (int i = 0; i < parsed.size(); ++i) {
-            prevToken = thisToken;
-            thisToken = parsed.get(i);
+    /**
+     * Adds a new operator to the top of the stack. If the top of the stack contains any operations with a priority
+     * higher than or equal to this operator, they are evaluated first.
+     *
+     * @return True on success, false on failure.
+     */
+    private static boolean handleOperator(@NotNull List<StackElement> stack, Operator op, Context ctx) {
+        if (stack.isEmpty()) {
+            ctx.success = false;
+            ctx.errorMessage = "Syntax error: no left-hand value for operator " + op.name;
+            return false;
+        }
+        if (stack.get(stack.size() - 1).isOperator) {
+            ctx.success = false;
+            ctx.errorMessage = "Syntax error: two operators in a row: " + stack.get(stack.size() - 1).operator.name
+                    + ", "
+                    + op.name;
+            return false;
+        }
+        // Evaluate any preceding operations with equal or higher priority than op.
+        // Exponentiation is right-associative, so in a ^ b ^ c we do not evaluate a ^ b yet.
+        evaluateStack(stack, op == Operator.POWER ? op.priority + 1 : op.priority);
+
+        stack.add(new StackElement(op));
+        return true;
+    }
 
-            if (prevToken == null && (thisToken instanceof Double || Operator.MINUS == thisToken)) continue;
-            if (prevToken instanceof Double && (thisToken instanceof Operator || thisToken instanceof Suffix)) continue;
-            if (prevToken instanceof Operator && thisToken instanceof Double) continue;
-            if (prevToken instanceof Suffix && (thisToken instanceof Operator || thisToken instanceof Suffix)) continue;
-            return DEFAULT;
+    /**
+     * Special handling for minus, we need to determine whether this is a unary or binary minus. If the top of the stack
+     * is a number, this is binary; if the stack is empty or the top is an operator, this is unary.
+     *
+     * @return True on success, false on failure.
+     */
+    private static boolean handleMinus(@NotNull List<StackElement> stack, Context ctx) {
+        if (stack.isEmpty() || stack.get(stack.size() - 1).isOperator) {
+            // Unary minus.
+            stack.add(new StackElement(0));
+            stack.add(new StackElement(Operator.UNARY_MINUS));
+        } else {
+            // Binary minus.
+            if (!handleOperator(stack, Operator.MINUS, ctx)) return false;
         }
-        if (thisToken instanceof Operator) return DEFAULT;
+        return true;
+    }
 
-        return parsed;
+    /**
+     * Handles adding a suffix on top of the stack. Suffixes are never actually added to the stack, since they have the
+     * highest priority the value on top of the stack is directly modified by the suffix.
```

Comment:
"they have the highest priority _and_ the value"?

Reply:
"Since a suffix has the highest priority, this means that the value on top of the stack can be directly modified (and the suffix never needs to live on the stack as its own thing)"

Reply:
I see. English is hard...